cpp/include/ucxx/delayed_submission.h

```diff
@@ -56,15 +57,135 @@ class DelayedSubmission {
                     const ucs_memory_type_t memoryType = UCS_MEMORY_TYPE_UNKNOWN);
 };
 
+template <typename T>
+class DelayedSubmissionCollectionTemplateSafe {
+ protected:
+  std::string_view _name{
+    "undefined"};                ///< The human-readable name of the collection, used for logging
+  std::vector<T> _collection{};  ///< The collection.
+  std::mutex _mutex{};           ///< Mutex to provide access to `_collection`.
+
+ public:
+  explicit DelayedSubmissionCollectionTemplateSafe(const std::string_view name) : _name{name} {}
+  DelayedSubmissionCollectionTemplateSafe()                                               = delete;
+  DelayedSubmissionCollectionTemplateSafe(const DelayedSubmissionCollectionTemplateSafe&) = delete;
+  DelayedSubmissionCollectionTemplateSafe& operator=(
+    DelayedSubmissionCollectionTemplateSafe const&)                                    = delete;
+  DelayedSubmissionCollectionTemplateSafe(DelayedSubmissionCollectionTemplateSafe&& o) = delete;
+  DelayedSubmissionCollectionTemplateSafe& operator=(DelayedSubmissionCollectionTemplateSafe&& o) =
+    delete;
+
+  /**
+   * @brief Register a callable or complex-type for delayed submission.
+   *
+   * Register a simple callback, or complex-type with a callback (requires specialization),
+   * for delayed submission that will be executed when the request is in fact submitted when
+   * `process()` is called.
+   *
+   * @throws std::runtime_error if delayed request submission was disabled at construction.
+   *
+   * @param[in] item      the callback that will be executed by `process()` when the
+   *                      operation is submitted.
```

Comment:
<!-- Nitpicks are trivial preference-based requests. These should be non-blocking by nature.  -->
**nitpick (non-blocking):** ‏Document `contextInitialized`?

Reply:
[Done](https://github.com/rapidsai/ucxx/pull/89/commits/c08cf4976cbc4a68e158eeeb15184550b1a0fa9f) along with some improvements.